### ISSUE TITLE
Xbox Gamepad controller not working in hardened security sandbox

### ISSUE THREAD

**Issue by bwayan** [sandboxie-plus/Sandboxie#2278]
### Describe what you noticed and did

Hello guys

I just subscribed to the support and I wanted to activate security enforcement protection on my sandboxes for my games.
However, when I switch from "standard" (yellow color) to orange/red ones ... it is impossible to play with my controller anymore.

If I go back to standard or blue one, working again.

I do not know on what additional information I shall share.
Windows 11 64 bits
Sandboxie v1.3.2

I get this error in the logs

                                                                                                                             
SBIE2112 Non accessible object : \Device\Afd\Endpoint, call CreateFile (C0000022) access=0016019F initialized=1   

### Affected program

Any programs which requires the controlpad

### Download link

N/A

### Where is the program located?

Not relevant to my request.

### Expected behavior

N/A

### What is your Windows edition and version?

Windows 11 64 bit

### In which Windows account you have this problem?

I use the built-in Administrator account.

### Please mention any installed security software

N/A

### What version of Sandboxie are you running?

1.3.2

### Is it a new installation of Sandboxie?

I have been using the same version for some time.

### In which sandbox type you have this problem?

In a Hardened sandbox (red sandbox icon).

### Can you reproduce this problem on an empty sandbox?

I can confirm it also on an empty sandbox.

### Did you previously enable some security policy settings outside Sandboxie?

No

**Comment by Luro223** [comment]
@bwayan as you're using the hardened sandbox, you need to add Xbox Gamepad controller manually in `sandbox`

**Comment by bwayan** [comment]
Thanks @Luro223 

I tried to add IPC Path  \Device\Afd\Endpoint* but it did not work.
Do you have any reference on how to add a controller to hardened sandbox?

Thanks

**Comment by Simba98** [comment]
> Thanks @Luro223
> 
> I tried to add IPC Path \Device\Afd\Endpoint* but it did not work. Do you have any reference on how to add a controller to hardened sandbox?
> 
> Thanks

 \Device\Afd\Endpoint* is for network, so it may not solve the Gamepad issue.

I think the needed sandboxie config is related to if you connect the Xbox Gamepad via Bluetooth or USB cable.
Compare with normal sandbox (Yellow box), I think some IPC is all the yellow box needed.

And you may refer to use Trace Log to Debug it by yourself.
![image](https://user-images.githubusercontent.com/7941967/192417558-83b312e0-7aa3-421e-99fd-1bb69cd65c99.png)

**Comment by Luro223** [comment]
@bwayan Have you tried Kernel Mode Object Filtering?

**Comment by bwayan** [comment]
Hello @Simba98 

The gamepad is on USB only.
As requested I activated it, but there is no many logs (registry, etc) that I do not know exactly where to look at
I filtered on RtClass and found these:


|Type|    |Status|   |Value|                                                    |Count|   

RtClass              Windows.Gaming.Input.RawGameController                     1         
RtClass              Windows.ApplicationModel.Core.CoreApplication              2         
RtClass              Windows.Gaming.Input.Custom.GameControllerFactoryManager   2         
RtClass              Windows.Gaming.Input.Gamepad                               2         
RtClass              Windows.UI.Core.CoreWindow                                 2         

Key		\registry\machine\SYSTEM\CurrentControlSet\Control\MediaProperties\PrivateProperties\Joystick\OEM\VID_045E&PID_028E	1
Ipc		\GLOBAL??\USB#VID_045E&PID_028E#0F98F1A#{ec87f1e3-c13b-4100-b5f7-8b84d54260cb}	160

But I have no idea on what to do

**Comment by Luro223** [comment]
@bwayan I don't have similar issue with my USB gamepad so I don't know how I can reproduce the issue

**Comment by Luro223** [comment]
I think we need the same model of gamepad and same drivers as well

**Comment by Luro223** [comment]
You can try to post the gamepad's driver here, maybe Isaak will help you

**Comment by Luro223** [comment]
@bwayan how are the results?

**Comment by bwayan** [comment]
Hello @Luro223 

Sorry for the late reply, busy days at work :)
I have a Xbox 360 USB controller for PC Windows.

Connected to Port_#0003.Hub_#0002

![image](https://user-images.githubusercontent.com/88817441/193054044-b2acccec-916e-40e7-b962-8debfedbd7d6.png)

**Comment by Luro223** [comment]
@bwayan Try to enable monitoring mode for tracelog and specified with com, gui, ipc, reg-keys, pipe, file access checked and post tracelog here
![541](https://user-images.githubusercontent.com/42772886/193349382-595e4e57-8c99-4a46-9517-cf34fd7b527d.png)

**Comment by isaak654** [comment]
I think you could try to add `OpenIpcPath=\Device\USBPDO-*` in the sandbox section of your Sandboxie.ini file.

**Comment by Luro223** [comment]
@bwayan if your main driver for your XBOX gamepad is \Device\USBPDO-3 then add `OpenIpcPath=\Device\USBPDO-*` as isaak described above.

**Comment by bwayan** [comment]
Hello guys @Luro223 and @isaak654 

Added the line, but unfortunately it did not have any effect.

@Luro223 
I am sorry to sound dumb, but where to find the gamepad driver?
I have that info: 
xusb22.inf:db04a16ccd17f2b6:CC_Install:10.0.22000.653:USB\VID_045E&PID_028E

**Comment by Luro223** [comment]
@bwayan installation package for your gamepad driver.
You can post it here, I will unpack and check driver.

**Comment by Luro223** [comment]
@bwayan hmm, you're trying to use the win7 driver on win11 if your xbox controller works fine even though these drivers for win7 then try to add following:
`OpenIpcPath=*\USB\Vid_045E&Pid_028E*`

**Comment by bwayan** [comment]
Hello @Luro223 

Sorry again for the late reply; I was travelling.
I tried both beloz but it is still KO unfortunately.

OpenIpcPath=\GLOBAL??\USB#VID_045E*
if still does not work then add this next to above :
OpenIpcPath=\Device\USBPDO-* (that one was already active as per @isaak654 recommendation above)

Note: I did not use the installer above as the drivers are already part of Windows 10 /11 natively. I just shared "an installer" if you needed to find the same in there.

**Comment by Luro223** [comment]
@bwayan 
```
OpenIpcPath=*VID_045E*
OpenIpcPath=\Device\USBPDO-*
```

**Comment by bwayan** [comment]
Hello @Luro223 

sorry for the late reply, on travel again (now no more travels so I shall reply way faster)

Unfortunately your last proposal did not work.
Is there anything else I can do to give you clues?

**Comment by Luro223** [comment]
@bwayan I don't think I can help you in this case. I don't have the compatible xbox controller and I can't emulate the controller. I can't reproduce the issue in any way.
Try to do `OpenIpcPath=*` as a check, if this does not work, then there is no workaround for your case.

**Comment by bwayan** [comment]
Hello @Luro223 

Unfortunately, even that one does not work.
Would it mean it does not use IPC but something else?

**Comment by DavidXanatos** [comment]
There must be a workaround we only need to figure out what path to open, hardened boxes don't lock down IPC much more, but the general driver endpoint access,
try, instead of opting for hardened box, to use the individual options, syscall lockdown and driver/device restrictions.
I think driver/device restrictions is what breaks it, but you need to test it for us to know.

**Comment by isaak654** [comment]
@bwayan
In a nutshell, David's suggestion is to disable the "Enable all security enhancements" feature in the hardened sandbox and play with the remaining two options in the following screenshot:

![Security_options](https://user-images.githubusercontent.com/12372772/197574944-77c56725-d181-43c6-8b1b-6cfd74b92564.png)

**Comment by bwayan** [comment]
Hello guys
@isaak654 and @DavidXanatos 

Since first post:
- I have updated from 1.4.2 to 1.5.1
- I have updated Windows 11 to latest patch

Here are my findinds;
- IPC: Authorizing \* ==> No effect
- Disabling Enable all (cf screenshot above) but leaving 2 below checked ==> No effect (Sandbox moves from Hardened Red to Sandbox with Data protection BLUE) 
- .... buuuuuuuut, If I stay in BLUE (work also in YELLOW) and I only check the first one (second one unchecked) ==> then the xBOX controller is working again

![image](https://user-images.githubusercontent.com/88817441/198266275-efc2648c-0722-4579-b89c-b784623e309d.png)

**Comment by isaak654** [comment]
Good news, David just pushed a permanent fix scheduled for 1.5.3 / 5.60.3: https://github.com/sandboxie-plus/Sandboxie/commit/52abe8986ec4a175dd54028c7760e15deb080866

The workaround should no longer be necessary in the next build.

**Comment by DavidXanatos** [comment]
I don't think this fix will do aynthing for the xbox controller the SRP device sit for software restriction policies only.

To solve the controller issue please try this how to: https://youtu.be/4k2XoT7VQuc
find out which "\device\" path is used by your controller by trying setting the blocked devices to normal until the one is found which is required.

**Comment by isaak654** [comment]
It can also be reproduced on Windows 10 21H2 x64 as well.

The following lines work for me even after enabling the option `Restrict driver/device access to only approved ones`:
```
NormalFilePath=\Device\00000054
NormalFilePath=\Device\USBPDO-*
```

I had to discard a dozen other devices, so it would probably be better to improve this part in the Trace Log.

**Comment by bwayan** [comment]
OMG !!!

@isaak654 and @DavidXanatos 
Yes it it working with hardened sandbox !!!
You rock it.

Will I have to add it manually to all my sandboxes that require my Xbox controller?

**Comment by DavidXanatos** [comment]
Do you need \Device\00000054 or only \Device\USBPDO-*  ?

**Comment by isaak654** [comment]
They're both needed at the same time. 
About the first one, wouldn't it be possible to associate it in the Trace Log with the Hardware ID shown in Device manager?

**Comment by bwayan** [comment]
Hello @DavidXanatos and @isaak654 

on my PC:
```
NormalFilePath=\Device\00000054 ==> KO if alone
NormalFilePath=\Device\USBPDO-* ==> OK if alone: Xbox controller is working
```

**Comment by isaak654** [comment]
So in Windows 10 I need both, while in Windows 11 you only need the last one.
Just in case, are you sure that you already cleaned the previous attempts made here?

**Comment by isaak654** [comment]
So in Windows 10 I need both, while in Windows 11 you only need the last one.
Just in case, are you sure that you've already cleaned the previous attempts made here?

**Comment by bwayan** [comment]
> Just in case, are you sure that you've already cleaned the previous attempts made here?

Yes, I rollbacked everytime I was making a new change.

**Comment by bwayan** [comment]
@isaak654 and @DavidXanatos 

Thanks for your time on this.
Im closing this issue as it has been Id'd and to keep your github clean ;)

**Comment by bwayan** [comment]
Hello guys, @isaak654 and @DavidXanatos 

Sorry to be a mood killer but I changed my USB controller for a Bluetooth one, and the settings we found before are not working anymore.

I tried to find some workaround, but impossible ...

Btw, I am now running SBP v1.7.2
Rest is the same as above (Windows 11, etc ...)

**Comment by isaak654** [comment]
Unfortunately, I don't own a Bluetooth controller to test it out.

**Comment by bwayan** [comment]
> Unfortunately, I don't own a Bluetooth controller to test it out.

Is there a way I could search on my own like you found for the wired xbox controller?
I tried to run the logs with all settings on, but it is like searching for a needle... unless we know what specifics I should look at

Let me know on how I could help :)

**Comment by DavidXanatos** [comment]
I am not sure if that wil fork on all systems a so generic path sounds like when you change the USB port or move to an otehr PC it will have a different number.

**Comment by bwayan** [comment]
> I am not sure if that wil fork on all systems a so generic path sounds like when you change the USB port or move to an otehr PC it will have a different number.

You are right, the path "slightly" change everytime I disconnect / reconnect by Bluetooth pad.

After many attempts, this is the path that is working on a more generic way:
`NormalFilePath=\Device\00000???`

(moved from 6 zero to 5)

**Comment by NewKidOnTheBlock** [comment]
I've ran into the same problem.
I'm using an Xbox One controller @ Windows 10 22H2 64bit.

Once the option "Restrict driver/device access to only approved ones" is checked, my controller stops working.
So I've added these three lines to my box's config:
OpenIpcPath=\Device\USBPDO-*
NormalFilePath=\Device\00000054
NormalFilePath=\Device\USBPDO-*
But no avail :/

**Comment by NewKidOnTheBlock** [comment]
@bwayan 
You might be re-inventing the wheel here.
There already exists a template for Bluetooth:
https://github.com/sandboxie-plus/Sandboxie/issues/799#issuecomment-853983805
Question is why it doesn't work in orange boxes.

**Comment by bwayan** [comment]
Hello @NewKidOnTheBlock 

I only play in hardened box with secured data, hence I have to use these (new lines) in the config.ini:

NormalFilePath=\Device\000000d9
NormalFilePath=\Device\USBPDO-*

I am sure if it works in hardened (red box), it would work in orange box.

What you need to check (if it still does not work) are the PIP logs :)

**Comment by NewKidOnTheBlock** [comment]
Is there any method to find out the device ID in an easier manner?
Like in the Hardware Manager?
![deviceID](https://github.com/sandboxie-plus/Sandboxie/assets/12411527/5efe7eb9-939e-4ce1-8d07-3cdb6a43dff2)
I guess this is being covered by `NormalFilePath=\Device\USBPDO-*`
How do I find out the other ID?

**Comment by bwayan** [comment]
> Is there a less painful method to find out the device IDs? Like e.g. in the Hardware Manager? ![deviceID](https://user-images.githubusercontent.com/12411527/282290002-5efe7eb9-939e-4ce1-8d07-3cdb6a43dff2.png) I guess this is being covered by `NormalFilePath=\Device\USBPDO-*` How do I find out the other device ID?


To find the second line, you need to enable logs + filter by IPC (cf my screenshot)
Then you need to look for a specific line (starting with \device\ ...

For some programs, it is the same number coming along, for others it varies and you look to find the lowest value
(ex: here it would be 00000*)

![01](https://github.com/sandboxie-plus/Sandboxie/assets/88817441/4cc2ac43-973c-42d8-a2e5-898d8f943cc7)

==> NormalFilePath=\Device\00000*

**Comment by NewKidOnTheBlock** [comment]
The actual range is larger than I expected.
Today, the Xbox Controller had the ID `Device\00000060`

So this is my config now to make the USB-driven Xbox Controller work with an Orange Security Box:
```
NormalFilePath=\Device\USBPDO-*
NormalFilePath=\Device\000000*
```

**Comment by eadmaster** [comment]
i'm having a similar issue with Sandboxie Classic v5.69.9, but only with some games (e.g. the recently-released [Iron Meat](https://store.steampowered.com/app/1157740/Iron_Meat/)).

where can i add these?
````
NormalFilePath=\Device\0000007*
NormalFilePath=\Device\USBPDO-*
````

**Comment by NewKidOnTheBlock** [comment]
That's correct, if you use DefaultBox for your games.

**Comment by ihaddy** [comment]
is there any update on this? would love to just whitelist the controller but finding it even with logs is like finding a needle in a haystack like others have said, i can't find any direct mapping between windows device manager IDs for the controller and how sandboxie identifies it?

Just wanted to edit to say it's only with specific games (so far two unity games i've tried with) that my controller can't be found. I've tried wired controllers, wireless, all xbox, some non xbox, i've tried  even using a yellow box and even a white box with wide open permissions to narrow it down, with and without network permissions, and no controllers are found.

```Enabled=y
;BlockNetworkFiles=y
RecoverFolder=%Desktop%
RecoverFolder=%Personal%
RecoverFolder=%{374DE290-123F-4565-9164-39C4925E467B}%
BorderColor=#02f6f6,ttl
Template=AutoRecoverIgnore
Template=LingerPrograms
Template=BlockPorts
Template=qWave
Template=FileCopy
Template=SkipHook
Template=OpenBluetooth
ConfigLevel=10
UseFileDeleteV2=y
UseRegDeleteV2=y
AutoRecover=y
OpenIpcPath=*
OpenPipePath=*
OpenFilePath=*
CallTrace=*
FileTrace=*
PipeTrace=*
KeyTrace=*
IpcTrace=*
GuiTrace=*
ClsidTrace=*
NetFwTrace=*
DnsTrace=y
ApiTrace=y
HookTrace=y
DebugTrace=y
ErrorTrace=y
ReadIpcPath=$:*
```

After some searching it seems like it's because unity might not use xinput in some cases or something? so i try to force it by editing unit config files for the game, and still no luck with it detecting my controller when sandboxed. The controller does work non-sandboxed immediately.

**Comment by CookiePLMonster** [comment]
> After some searching it seems like it's because unity might not use xinput in some cases or something?

To my knowledge, Unity uses Raw Input directly, so yes it may be related.

Leaving this comment as I also just encountered the very same issue with an Unity game.

**Comment by NewKidOnTheBlock** [comment]
Trying to use a controller has become a game of chance.
It works only 1 out of 3 times, despite adding the following lines:
```
NormalFilePath=\Device\USBPDO-*
NormalFilePath=\Device\000000*
```
I noted that it's failing especially often in Unreal Engine games.

Could you bake-in better Controller support?
Right now, the necessary knowledge to make it work (sometimes) is well hidden in this three-year-old issue.

**Comment by NewKidOnTheBlock** [comment]
I just learned that it should be openIPCpath like this:

```
OpenIpcPath=\Device\USBPDO-*
OpenIpcPath=\Device\000000*
```

**Comment by NewKidOnTheBlock** [comment]
I'm trying to play Aaero2 (should be based on Unity), but despite using the OpenIPCpath listed above, the controller doesn't work.
This game can only be played with a controller.
